pkg/controller/core.oam.dev/v1alpha2/application/dispatch/gc.go

```diff
@@ -0,0 +1,95 @@
+/*
+Copyright 2021 The KubeVela Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dispatch
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	"github.com/oam-dev/kubevela/apis/core.oam.dev/v1beta1"
+)
+
+// GarbageCollector do GC according two resource trackers
+type GarbageCollector interface {
+	GarbageCollect(ctx context.Context, oldRT, newRT *v1beta1.ResourceTracker) error
+}
+
+// NewGCHandler create a GCHandler
+func NewGCHandler(c client.Client, ns string) *GCHandler {
+	return &GCHandler{c, ns, nil, nil}
+}
+
+// GCHandler implement GarbageCollector interface
+type GCHandler struct {
+	c         client.Client
+	namespace string
+
+	oldRT *v1beta1.ResourceTracker
+	newRT *v1beta1.ResourceTracker
+}
+
+// GarbageCollect delete the old resources that are no longer in the new resource tracker
+func (h *GCHandler) GarbageCollect(ctx context.Context, oldRT, newRT *v1beta1.ResourceTracker) error {
+	h.oldRT = oldRT
+	h.newRT = newRT
+	if err := h.validate(); err != nil {
+		return err
+	}
+	klog.InfoS("Garbage collect for application", "old", h.oldRT.Name, "new", h.newRT.Name)
+	for _, oldRsc := range h.oldRT.Status.TrackedResources {
+		isRemoved := true
+		for _, newRsc := range h.newRT.Status.TrackedResources {
+			if oldRsc.APIVersion == newRsc.APIVersion && oldRsc.Kind == newRsc.Kind &&
+				oldRsc.Namespace == newRsc.Namespace && oldRsc.Name == newRsc.Name {
+				isRemoved = false
+				break
+			}
+		}
+		if isRemoved {
+			toBeDeleted := &unstructured.Unstructured{}
+			toBeDeleted.SetAPIVersion(oldRsc.APIVersion)
+			toBeDeleted.SetKind(oldRsc.Kind)
+			toBeDeleted.SetNamespace(oldRsc.Namespace)
+			toBeDeleted.SetName(oldRsc.Name)
+			if err := h.c.Delete(ctx, toBeDeleted); err != nil {
+				klog.ErrorS(err, "failed delete a resource", "name", oldRsc.Name, "apiVersion", oldRsc.APIVersion, "kind", oldRsc.Kind)
+				return errors.Wrapf(err, "cannot delete resource %q", oldRsc)
+			}
+			klog.InfoS("Successfully GC a resource", "name", oldRsc.Name, "apiVersion", oldRsc.APIVersion, "kind", oldRsc.Kind)
+		}
+	}
+	// TODO is it necessary to delete the old resource tracker now?
```

Comment:
Of course, why not?

Reply:
In our previous talk, you mentioned that, whether it's possible to retrieve the history of app revisions' applied manifests. I think this may be an opportunity to do that. Leave the GCed resource tracker out of there and let its controller do something based on these historical resource trackers.

Reply:
That's not the same case. When resources are deleted, the resourceTracker left is meanless.

Reply:
Got it. Will modify to delete old resource tracker.